{"cells":[{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"_kg_hide-input":true,"_kg_hide-output":true},"cell_type":"code","source":["# This Python 3 environment comes with many helpful analytics libraries installed\n","# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python\n","# For example, here's several helpful packages to load in \n","\n","import numpy as np\n","import pandas as pd\n","import re\n","import plotly.offline as py\n","import plotly.graph_objects as go\n","from plotly.subplots import make_subplots\n","\n","import os\n","os.environ['QT_QPA_PLATFORM']='offscreen' # https://github.com/etetoolkit/ete/issues/296\n","\n","import matplotlib.pyplot as plt\n","%matplotlib inline\n","from colorama import Back, Style, Fore\n","\n","try:\n","    from Bio.Align.Applications import ClustalwCommandline\n","    from Bio import Align, pairwise2\n","except:\n","    !pip install biopython\n","    from Bio.Align.Applications import ClustalwCommandline\n","    from Bio import Align, pairwise2\n","try:\n","    from ete3 import Tree, TreeStyle\n","except:\n","    !pip install ete3\n","    from ete3 import Tree, TreeStyle"],"execution_count":null,"outputs":[]},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true,"_kg_hide-input":true,"_kg_hide-output":true},"cell_type":"code","source":["# covid-19 genome\n","with open(\"/kaggle/input/coronavirus-accession-sars-mers-cov2/sars.fasta\", \"r\") as file:\n","    sars_genome = file.read().split(\"genome\")[1].replace(\"\\n\", \"\")\n","# mers genome\n","with open (\"/kaggle/input/coronavirus-accession-sars-mers-cov2/mers.fasta\", \"r\") as file:\n","    mers_genome = file.read().split(\"genome\")[1].replace(\"\\n\", \"\")\n","# sars genome\n","with open (\"/kaggle/input/coronavirus-accession-sars-mers-cov2/cov2.fasta\", \"r\") as file:\n","    cov_genome = file.read().split(\"genome\")[1].replace(\"\\n\", \"\")\n","# civet with sars like cov genome\n","with open(\"/kaggle/input/ebolav-vs-sarscov-vs-mers/Civet-SARS.fasta\", \"r\") as file:\n","    civet_sl_cov_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\",\"\")\n","# bat with sars like cov genome\n","with open(\"/kaggle/input/ebolav-vs-sarscov-vs-mers/BAT-SARS.fasta\", \"r\") as file:\n","    bat_sl_cov_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\",\"\")\n","# ebola-5 genome\n","with open(\"/kaggle/input/ebolav-vs-sarscov-vs-mers/EBOLAV.fasta\", \"r\") as file:\n","    ebola5_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\",\"\")\n","# camel with cov genome\n","with open(\"/kaggle/input/cameluscovfasta/camelus.fasta\", \"r\") as file:\n","    camel_cov_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\", \"\")\n","# malaria genome\n","with open(\"/kaggle/input/cameluscovfasta/plasmodium-malariae.fasta\", \"r\") as file:\n","    malaria_genome = file.read().split(\"complete sequence\")[-1].replace(\"\\n\", \"\")\n","# hiv-2 genome\n","with open(\"/kaggle/input/cameluscovfasta/hiv2.fasta\", \"r\") as file:\n","    hiv2_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\", \"\")\n","# hedgehog with cov genome\n","with open(\"/kaggle/input/cameluscovfasta/hedgehog.fasta\", \"r\") as file:\n","    hedgehog_cov_genome = file.read().split(\"complete genome\")[-1].replace(\"\\n\", \"\")\n"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["lets have a look at a snippet of one of the genomes"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["hiv2_genome[:3000]"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["having a huge sequence and with everyting in single color could be difficult to read. So let's color-code these nucleotides in the sequences and have them represented as they are normally written."]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["# some helper functions for representation of sequences\n","\n","# usually DNA sequences are written in 10-nucleotide fashion\n","def ten_nucleotide_seq(genome):\n","    genes = []\n","    for ix, char in enumerate(genome):\n","        if ix != 0 and ix%10 == 0:\n","            genes.append(' ')\n","        genes.append(char)\n","    return ''.join(genes)\n","\n","# color code to represent genome sequences\n","nu_clr_switcher = {\n","    # standard color-codes\n","    'A': Back.GREEN,\n","    'C': Back.YELLOW,\n","    'G': Back.RED,\n","    'T': Back.BLUE,\n","    ' ': Style.RESET_ALL\n","}\n","protein_clr_switcher = {\n","    # color-code by proteinfamily's polarity\n","    'A': Back.BLUE,\n","    'V': Back.BLUE,\n","    'I': Back.BLUE,\n","    'L': Back.BLUE,\n","    'M': Back.BLUE,\n","    'F': Back.BLUE,\n","    'Y': Back.CYAN,\n","    'W': Back.BLUE,\n","    'H': Back.CYAN,\n","    'R': Back.RED,\n","    'K': Back.RED,\n","    'N': Back.GREEN,\n","    'Q': Back.GREEN,\n","    'E': Back.MAGENTA,\n","    'D': Back.MAGENTA,\n","    'S': Back.GREEN,\n","    'T': Back.GREEN,\n","    'G': Back.YELLOW,\n","    'P': Back.YELLOW,\n","    'C': Back.BLUE,\n","    ' ': Style.RESET_ALL\n","}\n","def seq_repr(genome_str, strand ='dna'):\n","    if strand == 'dna':\n","        genome_str = ten_nucleotide_seq(genome=genome_str)\n","        line_break_cntr = 0\n","        for i in range(len(genome_str)):\n","            if genome_str[i] == ' ':\n","                line_break_cntr += 1\n","                if line_break_cntr>0 and line_break_cntr%6==0:\n","                    text = \"\\n\"\n","                else:\n","                    text = nu_clr_switcher[genome_str[i]] + genome_str[i]\n","            else:\n","                text = nu_clr_switcher[genome_str[i]] + genome_str[i]\n","            print(text, end=\"\")\n","        Style.RESET_ALL\n","        \n","    if strand == 'protein':\n","        for i in range(len(genome_str)):\n","            if genome_str[i] in protein_clr_switcher:\n","                if genome_str[i] == 'S' and genome_str[i+1:i+4] == 'TOP':\n","                    text = Style.RESET_ALL + 'S'\n","                elif genome_str[i] == 'T' and genome_str[i-1] == 'S' and genome_str[i+1:i+3] == 'OP':\n","                    text = Style.RESET_ALL + 'T'\n","                elif genome_str[i] == 'P' and genome_str[i-3:i] == 'STO':\n","                    text = Style.RESET_ALL + 'P'\n","                else:\n","                    text = protein_clr_switcher[genome_str[i]] + genome_str[i]\n","            else:\n","                Style.RESET_ALL\n","                text = genome_str[i]\n","            print(text, end=\"\")"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["print(\"COVID-19 genome: \")\n","seq_repr(cov_genome)"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Comparison-of-composition-of-genomes\"></a>\n","# Comparison of composition of genomes\n","Comparing the genomic composition, helps Vaccinologist do [reverse vaccinology](https://en.wikipedia.org/wiki/Reverse_vaccinology) to discover candidate [antigents](https://en.wikipedia.org/wiki/Antigen) for vaccine development by analyzing the genome of a pathogen or a family of pathogens.\n","\n","For the comparison of composition of these genomes, I'll be looking at the<br>\n","- Oligonucleotide composition<br>\n","- GC content\n","\n","<a id=\"Oligonucleotide-composition\"></a>\n","### 1. Oligonucleotide composition:\n","[Oligonucleotides](https://en.wikipedia.org/wiki/Oligonucleotide) are DNA or RNA molecules made up of a sequence of nucleotides. The length of the oligonucleotide is represented as \"[k-mer](https://en.wikipedia.org/wiki/K-mer)\" ($k$ being the number of nucleotides in one oligonulceotide). In this analysis, I'll be looking at the **3-mers** (or **trinucleotides**) and **4-mer** (or **tetranucleotides**) compositions against their normalized frquency of occurrences."]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["# tri-nucleotide compositions\n","trimers = [\"AAA\", \"AAC\", \"AAG\", \"AAT\", \"ACA\", \"ACC\", \"ACG\", \"ACT\", \"AGA\", \"AGC\", \"AGG\", \"ATA\", \"ATC\", \"ATG\", \"CAA\", \n","           \"CAC\", \"CAG\", \"CCA\",\"CCC\",\"CCG\",\"CGA\",\"CGC\",\"CTA\",\"CTC\",\"GAA\",\"GAC\",\"GCA\",\"GCC\",\"GGA\",\"GTA\",\"TAA\",\"TCA\"]\n","\n","def trimer_composition(genome):\n","    trimer_dict = dict()\n","    for trimer in trimers:\n","        trimer_dict[trimer] = genome.count(trimer)\n","    return trimer_dict\n","\n","# tetra-nucleotide compositions\n","tetramers = ['AAAA','AAAC','AAAG','AAAT','AACA','AACC','AACG','AACT','AAGA','AAGC','AAGG','AAGT','AATA','AATC',\n","             'AATG','AATT','ACAA','ACAC','ACAG','ACAT','ACCA','ACCC','ACCG','ACCT','ACGA','ACGC','ACGG','ACGT',\n","             'ACTA','ACTC','ACTG','AGAA','AGAC','AGAG','AGAT','AGCA','AGCC','AGCG','AGCT','AGGA','AGGC','AGGG',\n","             'AGTA','AGTC','AGTG','ATAA','ATAC','ATAG','ATAT','ATCA','ATCC','ATCG','ATGA','ATGC','ATGG','ATTA',\n","             'ATTC','ATTG','CAAA','CAAC','CAAG','CACA','CACC','CACG','CAGA','CAGC','CAGG','CATA','CATC','CATG',\n","             'CCAA','CCAC','CCAG','CCCA','CCCC','CCCG','CCGA','CCGC','CCGG','CCTA','CCTC','CGAA','CGAC','CGAG',\n","             'CGCA','CGCC','CGCG','CGGA','CGGC','CGTA','CGTC','CTAA','CTAC','CTAG','CTCA','CTCC','CTGA','CTGC',\n","             'CTTA','CTTC','GAAA','GAAC','GACA','GACC','GAGA','GAGC','GATA','GATC','GCAA','GCAC','GCCA','GCCC',\n","             'GCGA','GCGC','GCTA','GGAA','GGAC','GGCA','GGCC','GGGA','GGTA','GTAA','GTAC','GTCA','GTGA','GTTA',\n","             'TAAA','TACA','TAGA','TATA','TCAA','TCCA','TCGA','TGAA','TGCA','TTAA']\n","\n","def tetramer_composition(genome):\n","    tetramer_dict = dict()\n","    for tetramer in tetramers:\n","        if tetramer in genome:\n","            tetramer_dict[tetramer] = genome.count(tetramer)\n","        else:\n","            tetramer_dict[tetramer] = 0\n","    return tetramer_dict"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["labels = [\"SARS\", \"MERS\", \"COVID-19\", \"Civet_SL_CoV\",\"Bat_SL_CoV\",\"Ebola\",\"Camel_CoV\",\"Malaria\",\"HIV\",\"Hedgehog_CoV\"]\n","\n","traces = []\n","for i, genome in enumerate((sars_genome,mers_genome,cov_genome,civet_sl_cov_genome,bat_sl_cov_genome,ebola5_genome,\n","                            camel_cov_genome,malaria_genome,hiv2_genome,hedgehog_cov_genome)):\n","    composition = trimer_composition(genome)\n","    total_composition = sum(composition.values())\n","    norm_freq = [count/total_composition for count in composition.values()]\n","    traces.append(go.Scatter(x=trimers,y=norm_freq,mode='lines',name=labels[i]))\n","    \n","layout=go.Layout(\n","    title=\"Trinucleotide Composition of genomes\",\n","    xaxis_title=\"Trinucleotides\",\n","    yaxis_title=\"Normalized frequencies\",\n","    width=800, height=600\n",")\n","\n","fig = go.Figure(data=traces, layout=layout)\n","py.iplot(fig, filename=\"3mer-composition-frequency\")\n"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["Now lets have a look at the tetramer composition of these genomes"]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":["labels = [\"SARS\", \"MERS\", \"COVID-19\", \"Civet_SL_CoV\",\"Bat_SL_CoV\",\"Ebola\",\"Camel_CoV\",\"Malaria\",\"HIV\",\"Hedgehog_CoV\"]\n","\n","traces = []\n","for i, genome in enumerate((sars_genome,mers_genome,cov_genome,civet_sl_cov_genome,bat_sl_cov_genome,ebola5_genome,\n","                            camel_cov_genome,malaria_genome,hiv2_genome,hedgehog_cov_genome)):\n","    composition = tetramer_composition(genome)\n","    total_composition = sum(composition.values())\n","    norm_freq = [count/total_composition for count in composition.values()]\n","    traces.append(go.Scatter(x=tetramers,y=norm_freq,mode='lines',name=labels[i]))\n","    \n","layout=go.Layout(\n","    title=\"Tetranucleotides Composition of genomes\",\n","    xaxis_title=\"Tetranucleotides\",\n","    yaxis_title=\"Normalized frequencies\",\n","    width=1000, height=600\n",")\n","\n","fig = go.Figure(data=traces, layout=layout)\n","py.iplot(fig, filename=\"4mer-composition-frequency\")"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["Well that's alot to put in one plot, but the good thing with **plotly** is that you can always turn off a few legends to compare the values just among the selected few.\n","\n","For a moment, lets just look at the tetranucleotide composition of all the \"corona-virus infected species\"."]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["labels = [\"SARS\", \"MERS\", \"COVID-19\", \"Civet_SL_CoV\",\"Bat_SL_CoV\",\"Camel_CoV\",\"Hedgehog_CoV\"]\n","\n","traces = []\n","for i, genome in enumerate((sars_genome,mers_genome,cov_genome,civet_sl_cov_genome,\n","                            bat_sl_cov_genome,camel_cov_genome,hedgehog_cov_genome)):\n","    composition = tetramer_composition(genome)\n","    total_composition = sum(composition.values())\n","    norm_freq = [count/total_composition for count in composition.values()]\n","    traces.append(go.Scatter(x=tetramers,y=norm_freq,mode='lines',name=labels[i]))\n","    \n","layout=go.Layout(\n","    title=\"Tetranucleotides Composition of genomes\",\n","    xaxis_title=\"Tetranucleotides\",\n","    yaxis_title=\"Normalized frequencies\",\n","    width=1000, height=600\n",")\n","\n","fig = go.Figure(data=traces, layout=layout)\n","py.iplot(fig, filename=\"4mer-composition-frequency\")"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"GC-Content\"></a>\n","### 2. GC Content:\n","**[GC content](https://en.wikipedia.org/wiki/GC-content)** (or **Guanine-Cytosine content**) is the percentage of nucleotides bases in a DNA/RNA molecule that are either guanine (G) or cytosine (C). GC content is always expressed as a percentage value and is supposed to remaine the same among genomes of same specie.\n","\n","$$\\frac{G + C}{A + C + G + T} \\times 100\\ \\%$$"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["def gc_content(genome):\n","    g_content = genome.count(\"G\")\n","    c_content = genome.count(\"C\")\n","    a_content = genome.count(\"A\")\n","    t_content = genome.count(\"T\")\n","    return 100*((g_content+c_content)/(a_content+c_content+g_content+t_content))\n","\n","gc_content_perc = [gc_content(sars_genome),gc_content(mers_genome),gc_content(cov_genome),gc_content(civet_sl_cov_genome),\n","                   gc_content(bat_sl_cov_genome),gc_content(ebola5_genome),gc_content(camel_cov_genome),\n","                   gc_content(malaria_genome),gc_content(hiv2_genome), gc_content(hedgehog_cov_genome)]\n","viruses = [\"SARS\", \"MERS\", \"COVID-19\", \"Civet_SL_CoV\",\"Bat_SL_CoV\",\"Ebola\",\"Camel_CoV\",\"Malaria\",\"HIV\",\"Hedgehog_CoV\"]\n","colors = [\"lightslategray\",] * 10\n","colors[2] = \"crimson\"\n","\n","fig = go.Figure(data=[go.Bar(\n","    x=viruses,\n","    y=gc_content_perc,\n","    text=gc_content_perc,\n","    marker_color=colors\n",")])\n","fig.update_traces(texttemplate=\"%{text:.2f}%\", textposition=\"outside\")\n","fig.update_layout(\n","    title_text=\"GC content distribution among genomes\",\n","    xaxis_title=\"Genomes\",\n","    yaxis_title=\"GC content (%)\"\n",")"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Proteins-Amino-Acids\"></a>\n","# Proteins & Amino Acids\n","The most commonly occuring amino acids are abbreviated by using 20 letters from the english alphabet (all letters except for B,J,O,U,X and Z). The protein strings are constructed from these 20 alphabets as symbols. Henceforth, a Protein Sequence can also be incorporated using the nucleotide sequences.\n","\n","<img src=\"https://1.bp.blogspot.com/-QpZ1VXFbJhE/V4d5EjT3FYI/AAAAAAAADvk/UGZK71q3KV4AwrOfR_bdEk2FnC93Cb-JwCLcB/s1600/codon.png\" width=550 height=400>\n","\n","The DNA codon table on [wiki](https://en.wikipedia.org/wiki/DNA_codon_table) dictates the details regarding the encoding of specific codons into amino acids alphabets.\n","\n"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["def gen_protein_seq(genome_str):\n","    # codon table from wiki\n","    protein = {\"TTT\" : \"F\", \"CTT\" : \"L\", \"ATT\" : \"I\", \"GTT\" : \"V\",\n","           \"TTC\" : \"F\", \"CTC\" : \"L\", \"ATC\" : \"I\", \"GTC\" : \"V\",\n","           \"TTA\" : \"L\", \"CTA\" : \"L\", \"ATA\" : \"I\", \"GTA\" : \"V\",\n","           \"TTG\" : \"L\", \"CTG\" : \"L\", \"ATG\" : \"M\", \"GTG\" : \"V\",\n","           \"TCT\" : \"S\", \"CCT\" : \"P\", \"ACT\" : \"T\", \"GCT\" : \"A\",\n","           \"TCC\" : \"S\", \"CCC\" : \"P\", \"ACC\" : \"T\", \"GCC\" : \"A\",\n","           \"TCA\" : \"S\", \"CCA\" : \"P\", \"ACA\" : \"T\", \"GCA\" : \"A\",\n","           \"TCG\" : \"S\", \"CCG\" : \"P\", \"ACG\" : \"T\", \"GCG\" : \"A\",\n","           \"TAT\" : \"Y\", \"CAT\" : \"H\", \"AAT\" : \"N\", \"GAT\" : \"D\",\n","           \"TAC\" : \"Y\", \"CAC\" : \"H\", \"AAC\" : \"N\", \"GAC\" : \"D\",\n","           \"TAA\" : \"STOP\", \"CAA\" : \"Q\", \"AAA\" : \"K\", \"GAA\" : \"E\",\n","           \"TAG\" : \"STOP\", \"CAG\" : \"Q\", \"AAG\" : \"K\", \"GAG\" : \"E\",\n","           \"TGT\" : \"C\", \"CGT\" : \"R\", \"AGT\" : \"S\", \"GGT\" : \"G\",\n","           \"TGC\" : \"C\", \"CGC\" : \"R\", \"AGC\" : \"S\", \"GGC\" : \"G\",\n","           \"TGA\" : \"STOP\", \"CGA\" : \"R\", \"AGA\" : \"R\", \"GGA\" : \"G\",\n","           \"TGG\" : \"W\", \"CGG\" : \"R\", \"AGG\" : \"R\", \"GGG\" : \"G\" \n","           }\n","    protein_seq = \"\"\n","    \n","    # generating the protein seq\n","    for i in range(0, len(genome_str)-(3+len(genome_str)%3), 3):\n","        protein_seq += protein[genome_str[i:i+3]]\n","    return protein_seq"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["lets have a look at the protein sequences of a few of these genome before we go for a comparative analysis"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["print(\"Protein Sequences in Malaria: \")\n","seq_repr(gen_protein_seq(malaria_genome), strand='protein')"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["print(\"Protein sequences in CoV: \")\n","seq_repr(gen_protein_seq(cov_genome), strand='protein')"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Protein-Strands-by-genome\"></a>\n","### 1. Protein Strands by genome"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["genomes = {\"SARS\": sars_genome,\"MERS\": mers_genome,\"COVID-19\": cov_genome,\"Civet_SL_CoV\": civet_sl_cov_genome,\n","           \"Bat_SL_CoV\": bat_sl_cov_genome,\"Ebola\": ebola5_genome,\"Camel_CoV\": camel_cov_genome,\n","           \"Malaria\": malaria_genome,\"HIV\": hiv2_genome,\"Hedgehog_CoV\": hedgehog_cov_genome}\n","\n","colors = [\"lightslategray\",] * 10\n","colors[2] = \"crimson\"\n","\n","# for virus_name, virus_gene in genomes.items():\n","#     print(virus_name, len(gen_protein_seq(virus_gene).split('STOP')))\n","\n","fig = go.Figure(data=[go.Bar(\n","    x=list(genomes.keys()),\n","    y=[len(gen_protein_seq(genome).split('STOP')) for genome in genomes.values()],\n","    text=[len(gen_protein_seq(genome).split('STOP')) for genome in genomes.values()],\n","    marker_color=colors\n",")])\n","\n","fig.update_traces(texttemplate=\"%{text:.2s}\", textposition=\"outside\")\n","fig.update_layout(\n","    title_text=\"Number of protein strands among genomes\",\n","    xaxis_title=\"Genomes\",\n","    yaxis_title=\"number of protein strands\"\n",")"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Amino-Acids-Distribution\"></a>\n","### 2. Amino Acids Distribution"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["def amino_acid_composition(protein_seq):\n","    amino_acids_dict = {'Y': 0, 'A': 0, 'Q': 0, 'D': 0, 'C': 0, 'G': 0, 'V': 0, 'T': 0, 'E': 0, 'N': 0, \n","                       'K': 0, 'R': 0, 'S': 0, 'I': 0, 'H': 0, 'M': 0, 'F': 0, 'L': 0, 'W': 0, 'P': 0}\n","    for amino_acid in amino_acids_dict:\n","        amino_acids_dict[amino_acid] = protein_seq.count(amino_acid)/len(protein_seq)*100\n","    return amino_acids_dict"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["fig = go.Figure()\n","for i in range(10):\n","    fig.add_trace(go.Bar(\n","        x=list(amino_acid_composition(gen_protein_seq(list(genomes.values())[i])).keys()),\n","        y=list(amino_acid_composition(gen_protein_seq(list(genomes.values())[i])).values()),\n","        name=list(genomes.keys())[i]\n","    ))\n","\n","# Here we modify the tickangle of the xaxis, resulting in rotated labels.\n","fig.update_layout(barmode=\"group\",\n","                  title=\"Distribution of Amino Acids by genomes\", \n","                  xaxis_title=\"Amino Acids\", \n","                  yaxis_title=\"Normalized Frequency (%)\")\n","fig.show()"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Finding-the-ORFs\"></a>\n","### 3. Finding the ORFs (Open Reading Frames)\n","The stretch of codons from start-codon (ATG) to stop-codon (TAA/TAG/TGA) is called [Open Reading Frame (ORF)](https://en.wikipedia.org/wiki/Open_reading_frame)\n"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["def get_orfs(protein_seq):\n","    orf_strands = []\n","    # generate ORFs\n","    for seq in protein_seq.split(\"STOP\"):\n","        for nu in range(len(seq)-1, -1, -1):\n","            if seq[nu] == \"M\":\n","                orf_strands.append(seq[nu:]+\"STOP\")\n","    # search in original seq\n","    patterns = \"|\".join(orf_strands)\n","    res_seq = re.sub(patterns, lambda x: x.group(0).lower(), protein_seq)\n","    # highlight the ORFs in the seq\n","    for nu in res_seq:\n","        if nu.isupper():\n","            print(Style.RESET_ALL + nu, end=\"\")\n","        else:\n","            print(Back.BLUE + nu.upper(), end=\"\")\n","    print(Style.RESET_ALL+\"\\n\")"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["the highlighted sequences represent the ORFs"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["for genome_name, genome in genomes.items():\n","    print(\"\\nORFs in {}'s protein sequences:\".format(genome_name))\n","    get_orfs(gen_protein_seq(genome))\n","    print()"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"summary-of-protein-seq\"></a>\n","### 4. Comparative Summary of Protein Sequences"]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":["genomes = {\"SARS\": sars_genome,\"MERS\": mers_genome,\"COVID-19\": cov_genome,\"Civet_SL_CoV\": civet_sl_cov_genome,\n","           \"Bat_SL_CoV\": bat_sl_cov_genome,\"Ebola\": ebola5_genome,\"Camel_CoV\": camel_cov_genome,\n","           \"Malaria\": malaria_genome,\"HIV\": hiv2_genome,\"Hedgehog_CoV\": hedgehog_cov_genome}\n","\n","colors = [\"lightslategray\",] * 10\n","colors[2] = \"crimson\"\n","\n","proteins_seq = dict()\n","for genome_name, genome in genomes.items():\n","    proteins_seq[genome_name] = gen_protein_seq(genome)\n","    \n","proteins_seq_lengths = dict()\n","for genome_name, protein_seq in proteins_seq.items():\n","    len_of_pro_seq = []\n","    for seq in protein_seq.split('STOP'):\n","        len_of_pro_seq.append(len(seq)+1) # adding 1 for including STOP codon in total len of protein seq\n","    proteins_seq_lengths[genome_name] = len_of_pro_seq\n","\n","fig = make_subplots(\n","    rows=2, cols=2,\n","    specs=[[{}, {}],\n","           [{\"colspan\": 2}, None]],\n","    subplot_titles=(\"Length of Shortest Protein Seq\", \"Length of Longest Protein Seq\", \"Average Length of Protein Seq\"))\n","\n","fig.add_trace(\n","    go.Bar(\n","    x=list(proteins_seq_lengths.keys()),\n","    y=[min(_len) for _len in proteins_seq_lengths.values()],\n","    text=[min(_len) for _len in proteins_seq_lengths.values()],\n","    texttemplate=\"%{text:.1s}\", textposition=\"outside\",\n","    marker_color=colors),\n","    row=1, col=1\n",")\n","\n","fig.add_trace(\n","    go.Bar(\n","    x=list(proteins_seq_lengths.keys()),\n","    y=[max(_len) for _len in proteins_seq_lengths.values()],\n","    text=[max(_len) for _len in proteins_seq_lengths.values()],\n","    texttemplate=\"%{text:.1s}\", textposition=\"outside\",\n","    marker_color=colors),\n","    row=1, col=2\n",")\n","\n","fig.add_trace(\n","    go.Bar(\n","    x=list(proteins_seq_lengths.keys()),\n","    y=[sum(lens)/len(lens) for lens in proteins_seq_lengths.values()],\n","    text=[sum(lens)/len(lens) for lens in proteins_seq_lengths.values()],\n","    texttemplate=\"%{text:.2f}\", textposition=\"outside\",\n","    marker_color=colors),\n","    row=2, col=1\n",")\n","\n","fig.update_layout(title_text=\"Comparative Summary of Protein Sequences\",\n","                  height=960, width=960, showlegend=False)\n","\n","fig.show()"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Sequence-Alignement\"></a>\n","# Sequence Alignement\n","The sequence alignment is a way of arranging the sequences of DNA/RNA or protein to identify regions of similarity that may be a consequence of functional, structural, or evolutionary relationships between the sequences. These in turn help in identifying the sequence similarities and developing homology models of protein structures. Alignments are often assumed to reflect a degree of evolutionary change between sequences descended from a common ancestor."]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["sars_genome_aligned = ClustalwCommandline('clustalw2', infile=sars_genome)\n","mers_genome_aligned = ClustalwCommandline('clustalw2', infile=mers_genome)\n","cov_genome_aligned = ClustalwCommandline('clustalw2', infile=cov_genome)\n","bat_genome_aligned = ClustalwCommandline('clustalw2', infile=bat_sl_cov_genome)\n","civet_genome_aligned = ClustalwCommandline('clustalw2', infile=civet_sl_cov_genome)\n","camel_genome_aligned = ClustalwCommandline('clustalw2', infile=camel_cov_genome)\n","hedgehog_genome_aligned = ClustalwCommandline('clustalw2', infile=hedgehog_cov_genome)\n","ebola_genome_aligned = ClustalwCommandline('clustalw2', infile=ebola5_genome)\n","malaria_genome_aligned = ClustalwCommandline('clustalw2', infile=malaria_genome)\n","hiv_genome_aligned = ClustalwCommandline('clustalw2', infile=hiv2_genome)"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["as the genomes sequences are too big in length, lets align and have a look at just the first few nucleodtides in the sequences"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["# comparing first 57 nucleotides' of all the genomes after aligning them\n","print(\"SARS:\\t \",end=\" \")\n","seq_repr(sars_genome_aligned.infile[:57])\n","print(Style.RESET_ALL + \"\\nMERS:\\t \",end=\" \")\n","seq_repr(mers_genome_aligned.infile[:57])\n","print(Style.RESET_ALL + \"\\nCOVID:\\t \",end=\" \")\n","seq_repr(cov_genome_aligned.infile[:57])\n","print(Style.RESET_ALL + \"\\nBAT:\\t \",end=\" \")\n","seq_repr(bat_genome_aligned.infile[:57])\n","print(Style.RESET_ALL + \"\\nCIVET:\\t \",end=\" \")\n","seq_repr(civet_genome_aligned.infile[:57])\n","print(Style.RESET_ALL + \"\\nCAMEL:\\t \",end=\" \")\n","seq_repr(camel_genome_aligned.infile[:57])\n","print(Style.RESET_ALL + \"\\nHEDGEHOG:\",end=\" \")\n","seq_repr(hedgehog_genome_aligned.infile[:57])\n","print(Style.RESET_ALL + \"\\nEBOLA:\\t \",end=\" \")\n","seq_repr(ebola_genome_aligned.infile[:57])\n","print(Style.RESET_ALL + \"\\nMALARIA: \",end=\" \")\n","seq_repr(malaria_genome_aligned.infile[:57])\n","print(Style.RESET_ALL + \"\\nHIV:\\t \",end=\" \")\n","seq_repr(hiv_genome_aligned.infile[:57])"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["let us also have a look at the first few **aligned protein sequences** of these genomes"]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":["sars_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(sars_genome))\n","mers_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(mers_genome))\n","cov_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(cov_genome))\n","bat_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(bat_sl_cov_genome))\n","civet_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(civet_sl_cov_genome))\n","camel_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(camel_cov_genome))\n","hedgehog_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(hedgehog_cov_genome))\n","ebola_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(ebola5_genome))\n","malaria_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(malaria_genome))\n","hiv_pro_aligned = ClustalwCommandline('clustalw2', infile=gen_protein_seq(hiv2_genome))\n","\n","# comparing first 50 nucleotides' of all the protein seq after aligning them\n","print(\"SARS Proteins:\\t  \",end=\" \")\n","seq_repr(sars_pro_aligned.infile[:50], strand='protein')\n","print(Style.RESET_ALL + \"\\nMERS Proteins: \\t  \",end=\" \")\n","seq_repr(mers_pro_aligned.infile[:50], strand='protein')\n","print(Style.RESET_ALL + \"\\nCOVID Proteins:   \",end=\" \")\n","seq_repr(cov_pro_aligned.infile[:50], strand='protein')\n","print(Style.RESET_ALL + \"\\nBAT Proteins: \\t  \",end=\" \")\n","seq_repr(bat_pro_aligned.infile[:50], strand='protein')\n","print(Style.RESET_ALL + \"\\nCIVET Proteins:   \",end=\" \")\n","seq_repr(civet_pro_aligned.infile[:50], strand='protein')\n","print(Style.RESET_ALL + \"\\nCAMEL Proteins:   \",end=\" \")\n","seq_repr(camel_pro_aligned.infile[:50], strand='protein')\n","print(Style.RESET_ALL + \"\\nHEDGEHOG Proteins:\",end=\" \")\n","seq_repr(hedgehog_pro_aligned.infile[:50], strand='protein')\n","print(Style.RESET_ALL + \"\\nEBOLA Proteins:\\t  \",end=\" \")\n","seq_repr(ebola_pro_aligned.infile[:50], strand='protein')\n","print(Style.RESET_ALL + \"\\nMALARIA Proteins: \",end=\" \")\n","seq_repr(malaria_pro_aligned.infile[:50], strand='protein')\n","print(Style.RESET_ALL + \"\\nHIV Proteins: \\t  \",end=\" \")\n","seq_repr(hiv_pro_aligned.infile[:50], strand='protein')"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Pairwise-Sequence-Alignment\"></a>\n","# Pairwise Sequence Alignment\n","earlier we had aligned every sequence based on the [Clustal](https://en.wikipedia.org/wiki/Clustal) algorithm, now let's align every other genome with the covid19 genomes."]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["print(\"Aligning the first 100 nucleotides of COVID-19 genome with every other genomes:\\n\\n\")\n","for genome_name, genome in genomes.items():\n","    if genome_name != 'COVID-19':\n","        print(\"** COVID-19 and {} genome **\".format(genome_name))\n","        alignments = pairwise2.align.globalxx(cov_genome[:100], genome[:100])\n","        print(pairwise2.format_alignment(*alignments[0], full_sequences=True))"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["Let us also align the Protein Sequences of these genomes. This sort of analysis is often used by Researchers to spot **Similarity in Protein Spikes**\n","\n","**[NOTE]** in order to avoid the confussion of misinterpretting the encoding of 'STOP' codon with the encoding for S, T and P amino acids, the STOP codons in the following protein seq alignment has been replaces by $'*'$ symbol!"]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":["print(\"Aligning the first 100 amino acids of COVID-19 genome's protein sequence with every other genomes's protein sequence:\\n\\n\")\n","\n","for genome_name, protein_seq in proteins_seq.items():\n","    if genome_name != 'COVID-19':\n","        print(\"COVID-19 and {} genomes' protein sequences\\n\".format(genome_name))\n","        alignments = pairwise2.align.globalxx(gen_protein_seq(cov_genome[:100]).replace('STOP','*'), protein_seq[:100].replace('STOP','*'))\n","        print(pairwise2.format_alignment(*alignments[0], full_sequences=True))"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["the `score` in all the above alignment is the similarity score base on the number of matches in nucleotides and the dashes that showed up upon alignment are what is known as `gaps`. It is these gaps that account for the genetic mutations occuring from the insertions or deletions in the sequence, sometimes also referred to as **indels**"]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Genome-Sequence-Similarity\"></a>\n","# Genome Sequence Similarity\n","**using Clustal algorithm**<br>\n","\n","for establishing the similarity among genomes, we'll be changing the scoring parameter<br>\n","for every match there will be a $+1$ point<br>\n","for every mismatch there will be a $-1$ point<br>\n","for every open gap there will be a $-0.5$ points<br>\n","for every extended gap there will be a $-0.1$ point<br>"]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["aligner = Align.PairwiseAligner()\n","aligner.mode = 'global'\n","aligner.match_score = 1\n","aligner.mismatch_score = -1\n","aligner.open_gap_score = -0.5\n","aligner.extend_gap_score = -0.1"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["print(\"Similarity scores between\\n\")\n","for genome_name, genome in genomes.items():\n","    if genome_name != 'COVID-19':\n","        score = aligner.score(cov_genome, genome)\n","        print(\"COVID-19 & {} genome sequences:\\t {} ({:.2f}%)\".format(genome_name, score,100*(score/len(cov_genome))))"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Generating-Phylogenetic-Tree\"></a>\n","# Generating a Phylogenetic Tree\n","A [Phylogenetic Tree](https://en.wikipedia.org/wiki/Phylogenetic_tree) is the tree of evolutionary relationship among various species based on their genetic characteristics.\n","### UPGMA (Unweighted Pair Group Method with Arithmetic mean) algorithm\n","This is where you can learn more about it [wiki](https://en.wikipedia.org/wiki/UPGMA)<br>\n","[NOTE] unlike the above sequence similarity, here only the mismatch cases are scored using the aligner and no scoring is done for gaps and extended gaps."]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["# set aligner to score on mismatches\n","aligner.match_score = 0\n","aligner.mismatch_score = 1\n","aligner.open_gap_score = 0\n","aligner.extend_gap_score = 0\n","\n","# code UPGMA by hand\n","def low_value_position(matrix):\n","    # cell value with min value\n","    minimum_value = float('inf')\n","    position_x = -1\n","    position_y = -1\n","\n","    # traverse the entire matrix  to find the smallest value\n","    for i in range(len(matrix)):\n","        for j in range(len(matrix[i])):\n","            if (matrix[i][j] < minimum_value):\n","                minimum_value = matrix[i][j]\n","                position_x = i\n","                position_y = j\n","    return position_x, position_y\n","\n","# Combine two labels into one and erase the position with less value\n","# params: List of labels, positions\n","def join_headers(headers, a, b):\n","    if b < a:\n","        a,b = b,a\n","\n","    # join labels in the first position\n","    headers[a] = \"(\" + headers[a] + \",\" + headers[b] + \")\"\n","\n","    # delete label from secnd position\n","    del headers[b]\n","\n","# Join two columns of the matrix and save the result (average) in the first column\n","# Params: matrix of differences, position of the columns\n","def join_matrix(matrix,a ,b):\n","    if b < a:\n","        a,b = b,a\n","\n","    # recalculate row for smaller index\n","    row = []\n","    for i in range(0,a):\n","        row.append((matrix[a][i] + matrix[b][i])/2)\n","    matrix[a] = row\n","\n","    # recalculate entire column\n","    for i in range(a+1, b):\n","        matrix[i][a] = (matrix[i][a] + matrix[b][i])/2\n","        \n","    # get the rest of the values in row i\n","    for i in range(b+1, len(matrix)):\n","        matrix[i][a] = (matrix[i][a] + matrix[i][b])/2\n","        # delete the second column\n","        del matrix[i][b]\n","\n","    # delete second row\n","    del matrix[b]\n","\n","def upgma(matrix, headers):\n","\n","    iteration = 0\n","    while len(headers) > 1:\n","#         print('\\niteration: ', iteration)\n","#         print('total clusters: ', len(matrix))\n","\n","        # locate the position in the matrix containing the lowest value\n","        x, y = low_value_position(matrix)\n","\n","        # update the headers\n","        join_headers(headers, x, y)\n","#         print('family grouping: ', headers)\n","\n","        # join two columns of the difference matrix in the position coordinates\n","        join_matrix(matrix,x,y)\n","        iteration += 1\n","\n","    # final result stored in the first position\n","    return headers[0]"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["headers = [\"COVID19\",\"SARS\",\"MERS\",\"Bat with SARS like CoV\",\"Civet with SARS like CoV\",\"Camel with CoV\",\"Hedgehog with CoV\",\"EBOLA\",\"Malaria\",\"HIV\"]\n","viruses = [cov_genome, sars_genome, mers_genome, bat_sl_cov_genome, civet_sl_cov_genome, camel_cov_genome, hedgehog_cov_genome, ebola5_genome, malaria_genome, hiv2_genome]\n","\n","difference_matrix = [\n","    [],\n","    [aligner.score(viruses[1], viruses[0])],# mers-sars\n","    [aligner.score(viruses[2], viruses[0]),aligner.score(viruses[2], viruses[1])],# cov-sars, cov-mers\n","    [aligner.score(viruses[3], viruses[0]),aligner.score(viruses[3], viruses[1]), aligner.score(viruses[3], viruses[2])],# civet-sars, civet-mers, civet-cov\n","    [aligner.score(viruses[4], viruses[0]),aligner.score(viruses[4], viruses[1]), aligner.score(viruses[4], viruses[2]), aligner.score(viruses[4], viruses[3])],# bat-sars, bat-mers, bat-cov, bat-civet\n","    [aligner.score(viruses[5], viruses[0]),aligner.score(viruses[5], viruses[1]), aligner.score(viruses[5], viruses[2]), aligner.score(viruses[5], viruses[3]), aligner.score(viruses[5], viruses[4])],\n","    [aligner.score(viruses[6], viruses[0]),aligner.score(viruses[6], viruses[1]), aligner.score(viruses[6], viruses[2]), aligner.score(viruses[6], viruses[3]), aligner.score(viruses[6], viruses[4]), aligner.score(viruses[6], viruses[5])],\n","    [aligner.score(viruses[7], viruses[0]),aligner.score(viruses[7], viruses[1]), aligner.score(viruses[7], viruses[2]), aligner.score(viruses[7], viruses[3]), aligner.score(viruses[7], viruses[4]), aligner.score(viruses[7], viruses[5]), aligner.score(viruses[7], viruses[6])],\n","    [aligner.score(viruses[8], viruses[0]),aligner.score(viruses[8], viruses[1]), aligner.score(viruses[8], viruses[2]), aligner.score(viruses[8], viruses[3]), aligner.score(viruses[8], viruses[4]), aligner.score(viruses[8], viruses[5]), aligner.score(viruses[8], viruses[6]), aligner.score(viruses[8], viruses[7])],\n","    [aligner.score(viruses[9], viruses[0]),aligner.score(viruses[9], viruses[1]), aligner.score(viruses[9], viruses[2]), aligner.score(viruses[9], viruses[3]), aligner.score(viruses[9], viruses[4]), aligner.score(viruses[9], viruses[5]), aligner.score(viruses[9], viruses[6]), aligner.score(viruses[9], viruses[7]), aligner.score(viruses[9],viruses[8])]\n","]\n","\n","phylogenetic_tree = upgma(difference_matrix, headers)\n","phylogenetic_tree"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["tree = Tree(phylogenetic_tree + \";\")\n","# render tree image\n","tree.render(\"%%inline\", w=500)"],"execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":["# show the phylogenetic family as circular tree\n","ts = TreeStyle()\n","ts.show_leaf_name = True\n","ts.mode = 'c'\n","ts.arc_start = 10 # 0 degrees = 3 o'clock\n","ts.arc_span = 180\n","\n","tree.render(\"%%inline\", w=800, tree_style=ts)"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Clustering-patients-genomes\"></a>\n","# Clustering all COVID-19 patients' genomes\n","using the same UPGMA Algorithm"]},{"metadata":{"trusted":true,"_kg_hide-input":true,"_kg_hide-output":false},"cell_type":"code","source":["with open(\"/kaggle/input/covid19-genomes/LC528232.txt\", \"r\") as file:\n","    LC528232 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/LC528233.txt\", \"r\") as file:\n","    LC528233 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/LC529905.txt\", \"r\") as file:\n","    LC529905 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/LC534418.txt\", \"r\") as file:\n","    LC534418 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/LC534419.txt\", \"r\") as file:\n","    LC534419 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/LR757995.txt\", \"r\") as file:\n","    LR757995 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/LR757996.txt\", \"r\") as file:\n","    LR757996 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/LR757998.txt\", \"r\") as file:\n","    LR757998 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN908947.txt\", \"r\") as file:\n","    MN908947 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN938384.txt\", \"r\") as file:\n","    MN938384 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN975262.txt\", \"r\") as file:\n","    MN975262 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN985325.txt\", \"r\") as file:\n","    MN985325 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN988668.txt\", \"r\") as file:\n","    MN988668 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN988669.txt\", \"r\") as file:\n","    MN988669 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN988713.txt\", \"r\") as file:\n","    MN988713 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN994467.txt\", \"r\") as file:\n","    MN994467 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN994468.txt\", \"r\") as file:\n","    MN994468 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN996527.txt\", \"r\") as file:\n","    MN996527 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN996529.txt\", \"r\") as file:\n","    MN996529 = file.read().replace(\"\\n\", \"\")\n","with open(\"/kaggle/input/covid19-genomes/MN996530.txt\", \"r\") as file:\n","    MN996530 = file.read().replace(\"\\n\", \"\")\n","    \n","patients = [\"COVID19\",\"LC528232\",\"LC528233\",\"LC529905\",\"LC534418\",\"LC534419\",\"LR757995\",\"LR757996\",\"LR757998\",\"MN908947\",\"MN938384\",\"MN975262\",\"MN985325\",\"MN988668\",\"MN988669\",\"MN988713\",\"MN994467\",\"MN994468\",\"MN996527\",\"MN996529\",\"MN996530\"]\n","patients_viruses = [cov_genome,LC528232,LC528233,LC529905,LC534418,LC534419,LR757995,LR757996,LR757998,MN908947,MN938384,MN975262,MN985325,MN988668,MN988669,MN988713,MN994467,MN994468,MN996527,MN996529,MN996530]\n","\n","diff_mat = []\n","for i in range(len(patients_viruses)):\n","    if i == 0:\n","        diff_mat.append([])\n","    else:\n","        row = []\n","        for j in range(0,i):\n","            row.append(aligner.score(patients_viruses[i],patients_viruses[j]))\n","        diff_mat.append(row)\n","\n","phylogenetic_tree = upgma(diff_mat, patients)\n","print(\"Phylogenetic Tree of the genomes of all COVID infected persons:\\n\", phylogenetic_tree)"],"execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":["print(\"Clustering the COVID-19 infected patients' genomes alongwith the initial COVID19 genome taken in our above analysis\")\n","tree = Tree(phylogenetic_tree + \";\")\n","ts = TreeStyle()\n","ts.show_leaf_name = True\n","ts.mode = 'c'\n","tree.render(\"%%inline\", w=800, tree_style=ts)"],"execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":["<a id=\"Conclusions\"></a>\n","# Conclusions\n","\n","- Based on the oligonucleotide compositions, all the genomes from coronaviridae family show similar composition (trinucleotide and tetranucleotide composition)\n","- Based on GC content, the composition of Hedgehog infected Corona virus is closely similar to COVID-19 genome's GC content.\n","- Judging by the number of protein strands, the count in COVID19's genome is very close to those found in Bat infected with SARS like corona virus.\n","- If you filter down some of the categories in the plotly-plot of distribution of amino acids, you'll see the percentage of amino acids in COVID-19's genome closely match to those of MERS, Civet infected with SARS like CoV, Camel with alpha CoV and Malaria except for some and low high contents among L (Leucine), P (Proline) and T (Threonine) amino acids.\n","- Checking by the generated Phylogenetic tree, the genomes of COVID-19 and malaria virus show the maximum similarity. Perhaps this also explains why there's a lot of buzz on the news and internet about trying [Hydroxycholorquine](https://en.wikipedia.org/wiki/Hydroxychloroquine) and [anti-retroviral drugs](https://www.deccanherald.com/city/top-bengaluru-stories/coronavirus-hiv-anti-retroviral-drugs-to-be-used-for-covid-19-treatment-814534.html) which happen to be the medicines for malaria and HIV viruses, respectively."]},{"metadata":{},"cell_type":"markdown","source":["I really hope this turns out to be of any help to any microbiologist or vaccinologist out there.\n","If you liked this kernel or found it useful, please consider upvoting this notebook.<br>Thank you. Stay Home & Stay Safe!"]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}